uint32 id_mod
obj[] adap
